### ISSUE TITLE
Option to close comments after X days or X comments

### ISSUE THREAD

**Issue by herbdool** [backdrop/backdrop-issues#3001]
## Describe your issue or idea

In Wordpress, out of the box, there's an option to disable commenting on an article after X number of days.

![screenshot from 2018-03-15 12-02-38](https://user-images.githubusercontent.com/1071483/37475361-ea1f53d4-2848-11e8-81ff-f50e9d4bf822.png)

Likewise, I've ported a contrib module, Comment Closer, from Drupal https://backdropcms.org/project/commentcloser. It wasn't well-used in Drupal, maybe because someone would have to install it separately. I can imagine, like in Wordpress, that if there was a basic option to close comments after a time span (or also after X number of comments, then it would be popular. It doesn't need to be as complicated as Comment Closer in order to be pretty useful for people.

**Comment by laryn** [comment]
Agreed that this feature is a good one and would support it in core.

Thanks for the port as well, glad to know that is an option for now. I wasn't aware a module already existed that did so and had a [similar thought to add it to the Scheduler module at some point](https://github.com/backdrop-contrib/scheduler/issues/7).

**Comment by quicksketch** [comment]
I also agree that automatically closing comments would be a good core feature. This is something I leverage whenever it's available, because almost all commenting happens at the time content is originally posted. After the initial first couple of weeks, most comments that come in months or years later tend to be spam (and then it becomes a lot of spam).

**Comment by herbdool** [comment]
I've got a simple solution ready for trying out.

**Comment by herbdool** [comment]
Just thought I should also have a checkbox so if it's unchecked that it ignores and hides the number of days.

**Comment by olafgrabienski** [comment]
> Just thought I should also have a checkbox ...

Yes, please! (Have had a look at the sandbox site. At the moment, It's still without the checkbox, right?)

One question: What is supposed to happen with existing content, when I change the value for "Close comments after [ ] days" on the content type settings? Will the existing content respect the new time span or the one that had been defined, when the content was created?

**Comment by herbdool** [comment]
No checkbox yet.

It will apply for all content of that content type even if created before enabling this.

**Comment by herbdool** [comment]
@laryn I pushed another change--there's a checkbox now. Not sure if I have to do something for it to show up in the test box or will it update automatically.

**Comment by laryn** [comment]
@herbdool I think if you close and reopen the PR it rebuilds the test box.

**Comment by laryn** [comment]
I tested the latest PR and it's looking pretty good! Here's what I did:

- Set comments to be closed after 1 day on Posts
- Opened comments on the default Post
- Added a comment to this Post
- Backdated the "Authored on" date to early March on this Post
- Ran cron and reloaded the Post -- comments were still visible but closed to new comments

**Comment by quicksketch** [comment]
On the weekly call, we noted that because this is a content-type-wide setting, you would run into a problem if you re-enabled comments manually after 2 weeks, the auto-closer would immediately re-close the comments after you just re-enabled them. We may need to track the comment closing setting on a per-node basis. Thoughts?

**Comment by herbdool** [comment]
Good point. If it was also a toggle per node it would mean in order to reopen comments someone would need to open comments *and* uncheck the autoclose toggle. That seems unintuitive. Maybe they still want it to autoclose but starting a countdown from the date they reopened.

**Comment by laryn** [comment]
So if I'm following, there would be a per-node-type setting and then each individual node would default to that type-wide setting, but with an option to override it?

**Comment by herbdool** [comment]
@laryn I think that'll work ok. I did some research and Disqus does something similar https://help.disqus.com/customer/en/portal/articles/466238-moderating-your-community#auto-closing

**Comment by herbdool** [comment]
@laryn @olafgrabienski this is ready for another review. I've added an override checkbox for the individual nodes so that if the auto closer is enabled for the content type, the override checkbox will become visible on the individual node so you can have it stay open.

I've also added tests and they're passing.

**Comment by laryn** [comment]
Tested as follows:

- Set Post type to allow comments and auto-close after 1 day
- Created a new post. _[Comments were open as expected]_
- Edited and back-dated the post. Ran cron. _[Comments closed automatically as expected]_
- Edited, opened comments and overrode comment closing. Ran cron. _[Comments remained open as expected]_
- Just for fun, opened and unchecked the override box, leaving the "Open" setting for comments. Ran cron. _[Comments closed automatically as expected]_

Looks good to me!

**Comment by herbdool** [comment]
@laryn thanks for thorough review!

@olafgrabienski thanks for the suggestion. I've updated the PR

**Comment by olafgrabienski** [comment]
@herbdool Thanks for implementing the time span display on individual nodes. I've tested it in the sandbox site: When I changed the value in the content type configuration, it was changed on the node as well. Very nice!

**Comment by quicksketch** [comment]
This looks great! Thanks everyone for your reviews here also! They are most helpful! I did a code review and found a few items that should be tidied up. The implementation approach looks sound and the newly added test looks perfect.

PR review: https://github.com/backdrop/backdrop/pull/2111#pullrequestreview-113454815

**Comment by herbdool** [comment]
@quicksketch I've added your suggestions. An unrelated test failed but everything else passed.

**Comment by quicksketch** [comment]
All changes look perfect.

I found one more thing though: https://github.com/backdrop/backdrop/pull/2111#pullrequestreview-115812543

**Comment by quicksketch** [comment]
The change was trivial so I updated the PR to include the recommended fix (copy the field schema into the comment update hook): https://github.com/backdrop/backdrop/pull/2111/commits/5c4a0b8ff7e160ae233477e15b3e651f51c597b5t

Back to needs review.

**Comment by herbdool** [comment]
Thanks! Thanks for the reminder too--I forgot about not using hook_schema_alter in the hook_update_N

So since it was just the small change I'm guessing I can review it? I took a look at @quicksketch's change. Looks good and I reopened the PR to spin up the site again. Tested and still works as expected.
